File: src/forager/ui/widgets/sidebar.py
from __future__ import annotations
from PySide6.QtCore import Qt, QSize, Signal
from PySide6.QtGui import QIcon
from PySide6.QtWidgets import (
    QWidget, QVBoxLayout, QLabel,
    QLineEdit, QListWidget, QListWidgetItem, QFrame,
)

from forager.core.game import Game
from forager.services.icon_provider import load_icon
from forager.ui.theme import C
from forager.ui.icons import load_icon as load_bundled_icon
from forager.ui import style
from forager.ui.widgets.download_box import DownloadBox

_SIDEBAR_W = 240


class Sidebar(QWidget):
    game_selected = Signal(object)
    search_changed = Signal(str)
    download_clicked = Signal()

    def __init__(self, parent=None):
        super().__init__(parent)
        self._games: list[Game] = []
        self._search_text = ""

        self.setFixedWidth(_SIDEBAR_W)
        self.setStyleSheet(style.surface(2))

        layout = QVBoxLayout(self)
        layout.setContentsMargins(12, 16, 12, 12)
        layout.setSpacing(8)

        self._build_search(layout)
        self._build_list(layout)
        self._build_download_box(layout)
        self._build_user_panel(layout)

    def _build_search(self, layout):
        self._search = QLineEdit()
        self._search.setPlaceholderText("Search games...")
        self._search.setClearButtonEnabled(True)
        self._search.setStyleSheet(style.lineedit_qss())
        self._search.textChanged.connect(self._on_search)
        layout.addWidget(self._search)

    def _on_search(self, text: str):
        self._search_text = text.strip().lower()
        self._rebuild_list()
        self.search_changed.emit(self._search_text)

    def _build_list(self, layout):
        self._list = QListWidget()
        self._list.setFocusPolicy(Qt.FocusPolicy.NoFocus)
        self._list.setHorizontalScrollBarPolicy(Qt.ScrollBarPolicy.ScrollBarAlwaysOff)
        self._list.setIconSize(QSize(22, 22))
        self._list.setSpacing(4)
        self._list.setStyleSheet(
            f"""
            QListWidget {{
                background: transparent; border: none; outline: none;
                padding-top: 4px; font-size: 12px;
            }}
            QListWidget::item {{
                padding: 3px 8px; border-radius: {C.RADIUS}px;
                color: {C.TEXT_MUTED};
            }}
            QListWidget::item:hover {{
                background-color: {C.COLOR_3}; color: {C.TEXT};
            }}
            QListWidget::item:selected {{
                background-color: rgba(102, 108, 255, 90);
                color: {C.ACCENT_2};
            }}
            """
        )
        self._list.itemSelectionChanged.connect(self._on_selection_changed)
        layout.addWidget(self._list, stretch=1)

    def _on_selection_changed(self):
        item = self._list.currentItem()
        if item is None:
            return
        game: Game = item.data(Qt.ItemDataRole.UserRole)
        if game is not None:
            self.game_selected.emit(game)

    def _on_double_clicked(self, item: QListWidgetItem):
        game: Game | None = item.data(Qt.ItemDataRole.UserRole)
        if game is not None:
            self.game_selected.emit(game)

    def _build_download_box(self, layout):
        self._download_box = DownloadBox()
        self._download_box.clicked.connect(self.download_clicked)
        layout.addWidget(self._download_box)

    def begin_download(self, name: str) -> None:
        self._download_box.begin(name)

    def set_download_progress(self, progress) -> None:
        self._download_box.set_progress(progress)

    def hide_download(self) -> None:
        self._download_box.hide_download()

    def _build_user_panel(self, layout):
        panel = QFrame()
        panel.setStyleSheet(style.surface_qss(3))
        panel_layout = QVBoxLayout(panel)
        panel_layout.setContentsMargins(10, 10, 10, 10)
        panel_layout.setSpacing(8)

        self._count_label = QLabel()
        style.label(self._count_label, C.TEXT_MUTED, size=12)
        panel_layout.addWidget(self._count_label)

        layout.addWidget(panel)

    def set_games(self, games: list[Game]):
        self._games = sorted(games, key=lambda g: (g.sort_key or g.name).lower())
        self._rebuild_list()

    def _rebuild_list(self):
        current = self._list.currentItem()
        keep = None
        if current is not None:
            keep = current.data(Qt.ItemDataRole.UserRole)

        self._list.blockSignals(True)
        self._list.clear()
        shown = 0
        for g in self._games:
            if self._search_text and self._search_text not in g.name.lower():
                continue
            item = QListWidgetItem()
            item.setText(g.name.replace("/", " / "))
            item.setData(Qt.ItemDataRole.UserRole, g)
            item.setToolTip(str(g.path))
            icon = load_icon(g, allow_network=False)
            if icon is not None:
                item.setIcon(QIcon(icon))
            else:
                item.setIcon(load_bundled_icon("box", C.TEXT_MUTED))
            self._list.addItem(item)
            if keep is not None and g == keep:
                self._list.setCurrentItem(item)
            shown += 1

        if keep is None and self._list.count():
            self._list.setCurrentRow(0)
        self._list.blockSignals(False)

        total = len(self._games)
        self._count_label.setText(f"{shown} of {total} games")

    def set_icon(self, game: Game, icon: QIcon):
        for i in range(self._list.count()):
            item = self._list.item(i)
            if item.data(Qt.ItemDataRole.UserRole) == game:
                item.setIcon(icon)
                return

    def select_game(self, game: Game):
        for i in range(self._list.count()):
            item = self._list.item(i)
            if item.data(Qt.ItemDataRole.UserRole) == game:
                self._list.setCurrentItem(item)
                return

    def focus_next(self, direction: int):
        row = self._list.currentRow() + direction
        if 0 <= row < self._list.count():
            self._list.setCurrentRow(row)
            return True
        return False

    def activate_current(self):
        item = self._list.currentItem()
        if item is not None:
            self._on_selection_changed()
            return True
        return False

File: src/ui/widgets/Sidebar.cpp
#include "ui/widgets/Sidebar.h"
#include "ui/widgets/DownloadBox.h"
#include "ui/Theme.h"
#include "ui/Style.h"
#include "ui/Icons.h"
#include "services/IconProvider.h"
#include "library/Metadata.h"

#include <QVBoxLayout>
#include <QLineEdit>
#include <QListWidget>
#include <QLabel>
#include <QFrame>
#include <QKeyEvent>
#include <algorithm>

using namespace theme;
using namespace theme::C;

Sidebar::Sidebar(QWidget* parent)
    : QWidget(parent)
{
    setFixedWidth(240);
    style::background(this, COLOR_2);

    auto* layout = new QVBoxLayout(this);
    layout->setContentsMargins(12, 16, 12, 12);
    layout->setSpacing(8);

    m_search = new QLineEdit(this);
    m_search->setPlaceholderText("Search games...");
    m_search->setClearButtonEnabled(true);
    m_search->setStyleSheet(style::lineeditQss());
    connect(m_search, &QLineEdit::textChanged, this, &Sidebar::onSearch);
    layout->addWidget(m_search);

    m_list = new QListWidget(this);
    m_list->setFocusPolicy(Qt::NoFocus);
    m_list->setHorizontalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
    m_list->setIconSize(QSize(22, 22));
    m_list->setSpacing(4);
    m_list->setStyleSheet(style::listQss());
    connect(m_list, &QListWidget::itemSelectionChanged, this, &Sidebar::onSelectionChanged);
    connect(m_list, &QListWidget::itemDoubleClicked, this, &Sidebar::onDoubleClicked);
    layout->addWidget(m_list, 1);

    m_downloadBox = new DownloadBox(this);
    connect(m_downloadBox, &DownloadBox::clicked, this, &Sidebar::downloadClicked);
    layout->addWidget(m_downloadBox);

    auto* panel = new QFrame(this);
    style::panel(panel, 3);
    auto* panelLayout = new QVBoxLayout(panel);
    panelLayout->setContentsMargins(10, 10, 10, 10);
    panelLayout->setSpacing(8);

    m_countLabel = new QLabel(panel);
    style::label(m_countLabel, TEXT_MUTED, 12);
    panelLayout->addWidget(m_countLabel);

    layout->addWidget(panel);
}

void Sidebar::onSearch(const QString& text) {
    m_searchText = text.trimmed().toLower();
    rebuildList();
    emit searchChanged(m_searchText);
}

void Sidebar::activateItem(QListWidgetItem* item) {
    if (!item) return;
    auto game = item->data(Qt::UserRole).value<Game>();
    if (!game.name().isEmpty())
        emit gameSelected(game);
}

void Sidebar::onSelectionChanged() {
    activateItem(m_list->currentItem());
}

void Sidebar::onDoubleClicked(QListWidgetItem* item) {
    activateItem(item);
}

void Sidebar::setGames(const QList<Game>& games) {
    m_games = games;
    std::sort(m_games.begin(), m_games.end(), [](const Game& a, const Game& b) {
        return metadata::sortKey(a).toLower() < metadata::sortKey(b).toLower();
    });
    rebuildList();
}

void Sidebar::rebuildList() {
    Game keep;
    bool hasKeep = false;
    auto* current = m_list->currentItem();
    if (current) {
        keep = current->data(Qt::UserRole).value<Game>();
        hasKeep = true;
    }

    m_list->blockSignals(true);
    m_list->clear();
    int shown = 0;
    for (const auto& g : m_games) {
        if (!m_searchText.isEmpty() && !metadata::matchesQuery(g, m_searchText))
            continue;
        auto* item = new QListWidgetItem();
        item->setText(QString(g.name()).replace("/", " / "));
        QVariant v;
        v.setValue(g);
        item->setData(Qt::UserRole, v);
        item->setToolTip(g.displayPath());
        QPixmap pix = iconprovider::loadIcon(g, false);
        if (!pix.isNull()) {
            item->setIcon(QIcon(pix));
        } else {
            item->setIcon(icons::loadIcon("box", TEXT_MUTED));
        }
        m_list->addItem(item);
        if (hasKeep && g == keep)
            m_list->setCurrentItem(item);
        ++shown;
    }

    if (!hasKeep && m_list->count())
        m_list->setCurrentRow(0);
    m_list->blockSignals(false);

    int total = m_games.size();
    m_countLabel->setText(QStringLiteral("%1 of %2 games").arg(shown).arg(total));
}

QListWidgetItem* Sidebar::findItem(const Game& game) const {
    for (int i = 0; i < m_list->count(); ++i) {
        auto* item = m_list->item(i);
        if (item->data(Qt::UserRole).value<Game>() == game)
            return item;
    }
    return nullptr;
}

void Sidebar::setGameArt(const Game& game, const QPixmap& pix) {
    if (auto* item = findItem(game))
        item->setIcon(QIcon(pix));
}

void Sidebar::selectGame(const Game& game) {
    if (auto* item = findItem(game))
        m_list->setCurrentItem(item);
}

bool Sidebar::focusNext(int direction) {
    int row = m_list->currentRow() + direction;
    if (row >= 0 && row < m_list->count()) {
        m_list->setCurrentRow(row);
        return true;
    }
    return false;
}

bool Sidebar::activateCurrent() {
    auto* item = m_list->currentItem();
    if (item) {
        onSelectionChanged();
        return true;
    }
    return false;
}

void Sidebar::beginDownload(const QString& name) {
    m_downloadBox->begin(name);
}

void Sidebar::setDownloadProgress(double percent, const QString& stage, double speed, double done, double total) {
    m_downloadBox->setProgress(percent, stage, speed, done, total);
}

void Sidebar::hideDownload() {
    m_downloadBox->hideDownload();
}
